### PR TITLE
Openfoam extend 3.2 update

### PR DESCRIPTION
**Pull request by wpoely86**
Add support for openfoam-extend 3.2

### CODE REVIEW

File: easybuild/easyblocks/o/openfoam.py [hunk 1]
```diff
@@ -282,15 +307,16 @@ def sanity_check_step(self):
                                                     "refineMesh", "vorticity"]]
         # check for the Pstream and scotchDecomp libraries, there must be a dummy one and an mpi one
         if 'extend' in self.name.lower():
-            libs = [os.path.join(libsdir, x, "libPstream.so") for x in ["dummy", "mpi"]] + \
-                   [os.path.join(libsdir, "libscotchDecomp.so")]
+            libs = [os.path.join(libsdir, "libscotchDecomp.so")]
+            if LooseVersion(self.version) < LooseVersion('3.2'):
+                libs.extend([os.path.join(libsdir, x, "libPstream.so") for x in ["dummy", "mpi"]])
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: hold on, did this actually disappear, or are you ignored something that went wrong during the build?

references?

File: easybuild/easyblocks/o/openfoam.py [hunk 2]
```diff
@@ -197,19 +213,28 @@ def configure_step(self):
...
+                }
+                dep_vars = [
+                    ("%(name)s_SYSTEM", "1"),
+                    ("%(name)s_DIR", "%(root)s"),
+                    ("%(name)s_BIN_DIR", "%(root)s/bin"),
+                    ("%(name)s_LIB_DIR", "%(root)s/lib"),
+                    ("%(name)s_INCLUDE_DIR", "%(root)s/include"),
+                ]
+                for var, val in dep_vars:
+                    env.setvar(var % cur_dep, val % cur_dep)
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: this is a bit weird, since you never need `%(name)s` in `val`, or `%(root)s` in `var`, so don't use `cur_dep`?

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: I cannot find a reference to it anymore in any `Makefile`. Test build works.

File: easybuild/easyblocks/o/openfoam.py [hunk 3]
```diff
@@ -83,24 +83,38 @@ def extract_step(self):
...
+                            shutil.move(source, target)
+                    os.chdir(openfoam_installdir)
             except OSError, err:
                 raise EasyBuildError("Failed to move all files to %s: %s", openfoam_installdir, err)
 
+    def patch_step(self):
+        """Adjust beginpath and start_dir to correct openfoam dir"""
+        beginpath = os.path.join(self.installdir, self.openfoamdir)
+        self.cfg['start_dir'] = beginpath
+        super(EB_OpenFOAM, self).patch_step(beginpath=beginpath)
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: why is this required? are you sure this is not going to break existing OpenFOAM easyconfigs?

testing will tell...

File: easybuild/easyblocks/o/openfoam.py [hunk 4]
```diff
@@ -83,24 +83,38 @@ def extract_step(self):
...
+                    os.rename(source, target)
+                else:
+                    mkdir(openfoam_installdir)
+                    for fil in contents_installdir:
+                        if fil != self.openfoamdir:
+                            source = os.path.join(self.installdir, fil)
+                            target = os.path.join(openfoam_installdir, fil)
+                            self.log.debug("Moving %s to %s" % (source, target))
+                            shutil.move(source, target)
+                    os.chdir(openfoam_installdir)
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: wrap this in `try:`, `except OSError as err:`

File: easybuild/easyblocks/o/openfoam.py (hunk 3 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: because the original unpacked dir is renamed

File: easybuild/easyblocks/o/openfoam.py (hunk 4 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: see one line below?

**Comment by boegel** [comment]
@wpoely86: any idea how this will blend with https://github.com/hpcugent/easybuild-easyblocks/pull/744 ?

File: easybuild/easyblocks/o/openfoam.py (hunk 2 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: if it makes you happy

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: does using it also work?

File: easybuild/easyblocks/o/openfoam.py (hunk 2 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: it's just too confusing as is imho

**Comment by wpoely86** [comment]
It will clash, hence my request to merge it first. I will merge develop in here then

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: Haven't run a complete test suite but as far as I can tell, it works

**Comment by boegel** [comment]
@wpoely86: there are a couple of minor style remarks there, but it works, so should go in soon

**Comment by boegel** [comment]
@wpoely86: please rebase on #744?

I'd like to use this to test https://github.com/hpcugent/easybuild-easyconfigs/pull/2048, https://github.com/hpcugent/easybuild-easyconfigs/pull/2023, https://github.com/hpcugent/easybuild-easyconfigs/pull/2216, https://github.com/hpcugent/easybuild-easyconfigs/pull/2142

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: I can't find anything on Pstream disappearing, and since this check is relatively recent (and not having the libPstream libraries went undetected for a while), cfr. https://github.com/hpcugent/easybuild-easyblocks/pull/436/files#diff-0edc74d66adc49c07871a687be0ef167R212, I'd like to get confirmation of this somehow...

I don't see any mention of Pstream in the OpenFOAM-Extend 3.2 release notes (http://sourceforge.net/p/openfoam-extend/foam-extend-3.1/ci/v3.2/tree/ReleaseNotes.txt), so it's fishy to me that this just disappeared...

Maybe @besserox has an idea?

Comment (besserox) [inline easybuild/easyblocks/o/openfoam.py]: The Pstream library has apparently been removed in this commit:
http://sourceforge.net/p/foam-extend/foam-extend-3.2/ci/4c23c3f9e812674bf54b6a26e58f32511bb6af69/

It looks like it is now integrated in the main foam library.

What is important to check is that OpenFoam executables and libraries (eg libfoam.so) are actually linked to the MPI library of the toolchain, and not the one from the ThirdParty directory that they provide.
Running `ldd lib/linux64GccDPOpt/libfoam.so` should help to check this.

**Comment by wpoely86** [comment]
@boegel It's ready for testing

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: This sounds like an additional sanity check command to me. @wpoely86?

File: easybuild/easyblocks/o/openfoam.py [hunk 5]
```diff
@@ -209,19 +225,26 @@ def configure_step(self):
         if LooseVersion(self.version) < LooseVersion("2") or openfoam_extend_v3:
             self.log.debug("List of deps: %s" % self.cfg.dependencies())
             for dep in self.cfg.dependencies():
-                self.cfg.update('prebuildopts', "%s_SYSTEM=1" % dep['name'].upper())
-                self.cfg.update('prebuildopts', "{0}_LIB_DIR=$EBROOT{0}/lib".format(dep['name'].upper()))
-                self.cfg.update('prebuildopts', "{0}_INCLUDE_DIR=$EBROOT{0}/include".format(dep['name'].upper()))
+                dep_name = dep['name'].upper(),
+                dep_root = get_software_root(dep['name'])
+                dep_vars = [
+                    ("%s_SYSTEM", "1"),
```
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: I knew I had a reason for using dicts. This currently doesn't work due to this line. I don't see what's so weird about not using every key in every string formatting. It's what dicts are for?

File: easybuild/easyblocks/o/openfoam.py (hunk 1 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: Will see if I can manage something but as I disable all third party stuff, this is extremely unlikely to happen.

Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: Doesn't hurt to have an extra sanity check for this imho, we've missed it before, we may introduce a bug that breaks it again...

**Comment by besserox** [comment]
BTW, thanks for working on this one. I will need it very soon. 
I hope it will make it in EasyBuild 2.5  :-)

**Comment by boegel** [comment]
@besserox: it's going to be tight, but still being considered

File: easybuild/easyblocks/o/openfoam.py (hunk 5 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: just hoist this out of the for loop below, and replace with:

```python
env.setvar('%s_SYSTEM' % dep_name, '1')
```

File: easybuild/easyblocks/o/openfoam.py (hunk 4 above)
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: ok

File: easybuild/easyblocks/o/openfoam.py [hunk 6]
```diff
@@ -128,16 +142,17 @@ def configure_step(self):
...
-                # OpenFOAM >= 3.0.0 can use 64 bit integers
-                if 'extend' not in self.name.lower() and LooseVersion(self.version) >= LooseVersion('3.0'):
-                    WM_env_var.append('WM_LABEL_SIZE') 
-                for env_var in WM_env_var:
-                    from_pat = r"^(setenv|export) (?P<var>%s)[ =](?P<val>.*)$" % env_var
-                    to_pat = r": ${\g<var>:=\g<val>}; export \g<var>"
-                    line = re.sub(from_pat, to_pat, line)
-                sys.stdout.write(line)
+            # disable any third party stuff, we use EB controlled builds
+            regex_subs = [(r"^(setenv|export) WM_THIRD_PARTY_USE_.*[ =].*$", r"# \g<0>")]
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: not so sure about this one... this is bound to break existing builds?

File: easybuild/easyblocks/o/openfoam.py [hunk 7]
```diff
@@ -169,12 +184,13 @@ def configure_step(self):
...
-                for comp_var, newval in comp_vars.items():
-                    line = re.sub(r"^(%s\s*=\s*).*$" % re.escape(comp_var), r"\1%s" % newval, line)
-                sys.stdout.write(line)
+            regex_subs = []
+            for comp_var, newval in comp_vars.items():
+                regex_subs.append((r"^(%s\s*=\s*).*$" % re.escape(comp_var), r"\1%s" % newval))
+            apply_regex_substitutions(fullpath, regex_subs)
 
         # enable verbose build for debug purposes
+        # starting with openfoam-extend 3.2, PS1 also needs to be set
```
Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: what's this about?

Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: In their almighty wisdom, the openfoam-extend devs decide to only print stuff when both `FOAM_VERBOSE` and `PS1` are set.

Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: so, we should define `$PS1` if it's not set then, to get verbose output when OpenFOAM is built as a job too?

not sure if that would have other fallout, so let's not do it in this PR please, since it's undergoing heavy testing currently to get it in for EB v2.5.0

Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: I tried but it gets unset somewhere (probably bash?). It's not a major issue. It already prints some output but some finer details are not printed.

Comment (boegel) [inline easybuild/easyblocks/o/openfoam.py]: OK, fine, at least we have *some* trace of it. :-)

Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: I might file a bug report about it :wink:

File: easybuild/easyblocks/o/openfoam.py (hunk 6 above)
Comment (wpoely86) [inline easybuild/easyblocks/o/openfoam.py]: no, only used in openfoam-extend 3.2 so far

**Comment by boegel** [comment]
Tested thoroughly with existing OpenFOAM(-Extend) easyconfigs + new OpenFOAM(-Extend) easyconfigs in https://github.com/hpcugent/easybuild-easyconfigs/pull/2023, https://github.com/hpcugent/easybuild-easyconfigs/pull/2048, https://github.com/hpcugent/easybuild-easyconfigs/pull/2142 and https://github.com/hpcugent/easybuild-easyconfigs/pull/2216.

So, good to go, thanks @wpoely86!
